#### Actualmente hay 741 municipios donde hubo incendio y solo tenemos 233 estaciones metorológicas. Por lo que si hacemos join de incencios con valores meteorológicos perderemos muchos datos sobre incendios. 
#### El objetivo es tratar de asignar la estacion meteorológica XEMA más cercana  a cada municicipio.

In [0]:
!pip install geopy tqdm

Requirement already satisfied: geopy in /databricks/python3/lib/python3.8/site-packages (2.3.0)
Requirement already satisfied: tqdm in /databricks/python3/lib/python3.8/site-packages (4.64.1)
Requirement already satisfied: geographiclib<3,>=1.52 in /databricks/python3/lib/python3.8/site-packages (from geopy) (2.0)
 WARNING: You are using pip version 21.0.1; however, version 22.3.1 is available.
You should consider upgrading via the '/databricks/python3/bin/python -m pip install --upgrade pip' command.

In [0]:
from functools import reduce
import pyspark.sql.functions as F
from geopy.geocoders import Nominatim
from pyspark.sql.window import Window
from pyspark.sql.functions import row_number
from pyspark.sql.types import StringType
import pandas as pd

In [0]:
#parquet
incendios_ocurridos_CATDF = spark.read.format("parquet") \
.load("/mnt/IncendiosForestalesCAT/raw/meteocat/incendios_forestales") \
.select(["codi_municipi","termemunic"]) \
.withColumnRenamed("termemunic","municipio") \
.withColumnRenamed("codi_municipi","codigo_municipio")

#parquet
alt_pend_riscDF = spark.read.format("parquet") \
.load("/mnt/IncendiosForestalesCAT/prep/alt_pend_risc") \
.select(["cod_idescat","municipio"]) \
.withColumnRenamed("cod_idescat","codigo_municipio")


In [0]:
incendios_ocurridos_CATDF.select(["municipio"]).distinct().count()

Out[4]: 741

In [0]:
alt_pend_riscDF.select(["municipio"]).distinct().count()

Out[5]: 946

In [0]:
# Obtener municipios que no tienen datos meteo
muni_incendios= incendios_ocurridos_CATDF.distinct()
muni_risc =alt_pend_riscDF.distinct()
missing_muni = muni_risc.alias("df1").join(muni_incendios.alias("df2"), F.col("df1.codigo_municipio")==F.col("df2.codigo_municipio"),"left")
missing_muni.filter(F.col("df2.municipio").isNull()).count()

Out[6]: 206

In [0]:
missing_muni=missing_muni.select("df1.*")
display(missing_muni)

codigo_municipio municipio 89019 Rupit i Pruit 431711 Vila-seca 250241 Alt Àneu 82074 Sant Esteve de Palautordera 80790 Estany, l' 80272 Cabanyes, les 82270 Sant Martí Sarroca 170524 Siurana 430976 Nou de Gaià, la 81115 Malla 171886 Selva de Mar, la 430939 Móra d'Ebre 170010 Agullana 251713 Pobla de Segur, la 80060 Arenys de Mar 80155 Badalona 431750 Vilalba dels Arcs 431172 Prat de Comte 250753 Clariana de Cardener 252094 Sort 81228 Mediona 252424 Verdú 251636 Coma i la Pedra, la 81712 Prats de Lluçanès 81173 Masies de Voltregà, les 430995 Palma d'Ebre, la 250485 Bell-lloc d'Urgell 82460 Santa Eugènia de Berga 250851 Estaràs 431515 Torre de Fontaubella, la 250693 Castelló de Farfanya 82227 Sant Llorenç d'Hortons 83054 Vilafranca del Penedès 259062 Valls d'Aguilar, les 170963 Llosses, les 252173 Tàrrega 82858 Torelló 80039 Alella 80522 Castellar de n'Hug 171772 Sant Pau de Segúries 80229 Berga 251410 Montoliu de Segarra 250946 Foradada 170932 Llers 430764 Marçà 82251 Sant Martí d'Albars 430786 Masdenverge 80636 Castellolí 81483 Olivella 170025 Aiguaviva 430402 Capçanes 171024 Maçanet de Cabrenys 252458 Vilaller 80689 Cervelló 171096 Montagut i Oix 82055 Sant Cugat del Vallès 431074 Pira 250899 Farrera 82554 Santa Maria de Merlès 80765 Esparreguera 82188 Sant Joan de Vilatorrada 170985 Maià de Montcal 250501 Bellpuig 250208 Almatret 431227 Renau 431378 Sant Jaume dels Domenys 80168 Bagà 80902 Gaià 251930 Sant Llorenç de Morunys 82763 Tagamanent 430290 Blancafort 251266 Llavorsí 179032 Sant Julià del Llor i Bonmatí 172039 Ultramort 80076 Arenys de Munt 251500 Oliola 430326 Bot 80291 Cabrera de Mar 82975 Veciana 430287 Bisbal del Penedès, la 171289 Pau 430509 Creixell 251898 Rosselló 170545 Colera 172095 Vall-llobrega 250349 Artesa de Segre 430897 Montferri 80253 Bruc, el 80945 Granada, la 170715 Fontcoberta 80713 Copons 430918 Mont-ral 250236 Alpicat 430957 Morell, el 431331 Roquetes 80023 Aguilar de Segarra 431188 Pratdip 250404 Balaguer 83036 Vilanova de Sau 251142 Ivorra 171674 Sant Joan de les Abadesses 171899 Cellera de Ter, la 82896 Torrelles de Llobregat 170911 Llanars 259010 Vall de Cardós 250292 Arbeca 171637 Sant Gregori 80608 Castellfollit de Riubregós 431573 Ulldemolins 250735 Cervià de les Garrigues 81189 Masnou, el 170078 Amer 81537 Òrrius 250962 Fuliola, la 252022 Senterada 431763 Vimbodí i Poblet 251057 Granyena de les Garrigues 171616 Sant Feliu de Pallerols 80556 Castellcir 250077 Albatàrrec 82037 Sant Cebrià de Vallalta 431359 Salomó 431299 Riudoms 251272 Lles de Cerdanya 430770 Mas de Barberans 431495 Tivenys 82900 Ullastrell 170274 Breda 170675 Flaçà 82457 Santa Coloma de Gramenet 82129 Sant Feliu Sasserra 431135 Pont d'Armentera, el 171169 Osor 251207 Lleida 172327 Vilopriu 81311 Montesquiu 170427 Capmany 82248 Sant Martí de Centelles 171236 Palol de Revardit 251832 Rialp 80018 Abrera 80517 Castellar del Vallès 171077 Molló 251961 Sant Esteve de la Sarga 80581 Castellet i la Gornal 81332 Montmaneu 82140 Vilassar de Dalt 170388 Campllong 251924 Sant Guim de Freixenet 80919 Gelida 430614 Forès 250450 Bausen 82981 Vic 170340 Calonge i Sant Antoni 430483 Corbera d'Ebre 430213 Barberà de la Conca 81120 Manlleu 431093 Pobla de Mafumet, la 172109 Ventalló 430748 Llorenç del Penedès 82956 Vallirana 170950 Lloret de Mar 170669 Figueres 431536 Torredembarra 81786 Rajadell 430825 Masroig, el 171292 Pedret i Marzà 252120 Sunyer 170681 Foixà 431628 Vandellòs i l'Hospitalet de l'Infant 179011 Cruïlles, Monells i Sant Sadurní de l'Heura 250169 Alguaire 80654 Castellví de la Marca 81509 Orís 250523 Bellvís 259118 Plans de Sió, els 81859 Rubió 250117 Alcarràs 431649 Vespella de Gaià 259031 Guingueta d'Àneu, la 81944 Sant Adrià de Besòs 81022 Igualada 81350 Montmeló 251538 Omellons, els 171518 Riumors 80109 Artés 170366 Campdevànol 171595 Sant Feliu de Buixalleu 81593 Parets del Vallès 81751 Puig-reig 431166 Prades 251749 Portella, la 80180 Balsareny 250560 Bovera 81939 Sant Iscle de Vallalta 430115 Almoster 430094 Alforja 8

Solo tenemos 1 municipio sin datos de cartografia, riesgo incendio y 206 municipios que no tenemos en la tabla incendio. 
Por lo tanto se analizará calculará la distancia entre municipios de alt_pend_riscDF y las estaciones AEMET y Meteocat

In [0]:
estaciones_meteocatDF = spark.read.parquet("/mnt/IncendiosForestalesCAT/raw/meteocat/estaciones/")\
.filter(F.col("data_fi") >= "nan")\
.select(["codi_estacio", "latitud", "longitud", "nom_municipi"])\
.withColumnRenamed("codi_estacio","indicativo")\
.withColumnRenamed("nom_municipi","nombre")
display(estaciones_meteocatDF)

indicativo latitud longitud nombre X4 41.3839 2.16775 Barcelona YG 42.51881 1.24244 Tírvia VY 41.25095 1.29863 Nulles YD 41.51135 0.85617 Les Borges Blanques VD 41.68939 1.20381 Els Plans de Sió WX 41.9178 0.88175 Camarasa XV 41.48311 2.07956 Sant Cugat del Vallès U1 42.30648 2.95481 Cabanes CY 41.87813 2.17873 Muntanyola D6 42.43515 3.16622 Portbou D3 41.38197 1.93564 Vallirana Z1 42.64691 0.98486 Alt Àneu X1 41.15374 0.81953 Falset VH 41.65814 0.39301 Gimenells i el Pla de la Font WK 41.81949 0.57768 Alfarràs YL 41.1227 0.96951 Riudecanyes CT 42.39811 0.74362 El Pont de Suert XR 41.31481 0.98161 Prades H1 41.58642 1.65306 Òdena UE 42.02306 3.15719 Torroella de Montgrí W4 41.36619 1.72857 La Granada X3 41.74281 0.5358 Alguaire US 40.55786 0.52329 Alcanar KP 41.73894 2.69436 Fogars de la Selva XD 41.32 0.8857 Ulldemolins XY 41.56509 0.55027 Alcarràs CJ 42.21624 1.33132 Organyà U9 40.76947 0.61635 L'Aldea YB 42.18813 2.47097 Olot D1 41.28521 0.75383 Margalef XG 41.56734 2.22619 Parets del Vallès UF 41.28832 1.90775 Begues Z3 42.46605 1.7785 Meranges CE 41.53109 1.80813 Els Hostalets de Pierola CU 42.69856 0.79397 Vielha e Mijaran WI 41.35741 0.4809 Maials UW 40.62725 0.65922 Sant Carles de la Ràpita CC 42.07398 2.20862 Orís W8 41.44237 1.15998 Blancafort C6 41.6566 0.95172 Castellnou de Seana WO 41.27151 1.46717 La Bisbal del Penedès VU 41.63286 1.91718 Rellinars WQ 42.0513 0.72952 Sant Esteve de la Sarga XQ 42.16252 0.88814 Tremp WB 41.76036 0.67022 Albesa DQ 41.30728 1.36259 Vila-rodona UI 42.26477 1.76218 Gisclareny DK 41.14677 1.41846 Torredembarra J5 42.34215 2.83423 Darnius Y9 41.31723 2.13091 El Prat de Llobregat WG 41.80104 0.64804 Algerri UJ 41.52879 1.3683 Santa Coloma de Queralt MV 42.13598 1.67295 Guixers VV 41.68129 2.02647 Sant Llorenç Savall UP 41.51773 2.37702 Cabrils XA 41.60257 1.4007 Montmaneu VC 41.24415 0.43266 Riba-roja d'Ebre Y4 42.17426 1.42775 Fígols i Alinyà Z5 42.70029 1.27201 Lladorre ZC 42.42117 2.24565 Setcases CL 41.67399 1.76796 Sant Salvador de Guardiola XT 41.98766 1.51165 Solsona X8 41.37919 2.1054 Barcelona UN 41.87449 2.92694 Cassà de la Selva WR 41.2163 0.79748 Torroja del Priorat Y5 42.22793 2.86295 Navata VE 41.48692 0.46058 Aitona UC 41.97726 2.99705 Cruïlles, Monells i Sant Sadurní de l'Heura DG 42.39848 2.15517 Queralbs XH 42.40533 1.12993 Sort V5 42.03947 2.11993 Perafita VN 41.88244 2.74262 Vilobí d'Onyar C7 41.66695 1.16234 Tàrrega YJ 41.58462 0.64172 Lleida VZ 42.38443 3.00736 Espolla WS 41.84008 2.41877 Viladrau CP 42.13924 1.03893 Isona i Conca Dellà CR 42.08032 1.9624 La Quar C9 40.71825 0.39988 Mas de Barberans VQ 41.1713 1.16774 Constantí CI 42.25839 2.36429 Sant Pau de Segúries U4 41.83094 1.84699 Castellnou de Bages MQ 41.92047 1.69584 Cardona DB 40.87287 0.71581 El Perelló D7 41.18499 0.59376 Vinebre WZ 41.20187 1.63346 Cunit WP 41.48518 1.69337 Mediona W5 42.07683 1.31489 Oliana UA 40.9091 0.76943 L'Ametlla de Mar Y7 41.31725 2.16537 El Prat de Llobregat X6 41.92173 1.02901 Artesa de Segre XE 41.10393 1.201 Tarragona XB 41.4779 1.53529 La Llacuna XZ 42.18139 3.06202 Torroella de Fluvià DO 41.80857 3.03241 Castell-Platja d'Aro V8 41.67279 0.87741 El Poal UB 42.05398 3.06195 La Tallada d'Empordà X7 41.51909 0.55314 Torres de Segre UK 41.27861 1.80479 Sant Pere de Ribes VB 41.06289 0.63517 Benissanet C8 41.67555 1.29609 Cervera WA 41.87694 1.1541 Oliola YK 41.55361 1.99005 Terrassa Z9 42.29265 1.71498 Montellà i Martinet VX 41.74761 2.30291 Tagamanent WV 42.23414 1.87485 Guardiola de Berguedà U2 42.17716 3.0968 Sant Pere Pescador D8 40.95134 0.30565 Horta de Sant Joan ZB 42.51939 1.36597 Alins WL 41.57236 1.0882 Sant Martí de Riucorb W9 42.14557 2.45451 La Vall d'en Bas WU 41.45215 2.24757 Badalona VO 42.08745 1.42792 Lladurs XO 41.93497 2.23987 Vic XC 41.47892 1.97546 Castellbisbal XK 41.77362 2.43774 Fogars de Montclús ZD 42.32211 1.89716 Alp WW 41.79423 1.9368 Artés Y6 41.04343 0.74032 Tivissa WD 41.08671 0.31884 Batea CQ 41.99546 1.02569 Vilanova de Meià XS 41.86489 2.664

In [0]:
@udf(returnType=StringType())
def convertToDec(coord_str):
    """
    Convertir coordenada con formato grados, minutos, segundos a decimales
    """
    direction = coord_str[-1]
    coord = (int(coord_str[:2]) +int(coord_str[2:4])/60 +float(coord_str[4:6])/3600)* (-1 if direction in ['W', 'S'] else 1)
    return (f"{coord}")

In [0]:
estaciones_aemetDF = spark.read.parquet("/mnt/IncendiosForestalesCAT/raw/aemet/estaciones/")\
.filter(F.col("provincia").isin(["BARCELONA", "TARRAGONA", "GIRONA", "LLEIDA"]))\
.select(["indicativo", "latitud", "longitud", "nombre"]) \
.withColumn("latitud", convertToDec(F.col("latitud"))) \
.withColumn("longitud", convertToDec(F.col("longitud")))
display(estaciones_aemetDF)

indicativo latitud longitud nombre 0252D 41.587500000000006 2.54 ARENYS DE MAR 0076 41.29277777777778 2.0700000000000003 BARCELONA AEROPUERTO 0200E 41.41833333333333 2.1241666666666665 BARCELONA, FABRA 0201D 41.39055555555556 2.2 BARCELONA 0149X 41.72 1.840277777777778 MANRESA 0229I 41.52361111111111 2.1030555555555557 SABADELL AEROPUERTO 0255B 41.65083333333333 2.6969444444444446 SANTA SUSANNA 0367 41.91166666666666 2.763333333333333 GIRONA AEROPUERTO 0370B 41.980000000000004 2.8252777777777776 GIRONA, ANTIC INSTITUT 0372C 42.10444444444445 2.763611111111111 PORQUERES 0324A 42.19138888888889 2.196111111111111 RIPOLL 0016A 41.144999999999996 1.163611111111111 REUS AEROPUERTO 0002I 40.95805555555556 0.8713888888888889 VANDELLÒS 9585 42.3325 1.9394444444444445 LA MOLINA 9619 42.35472222222222 1.4630555555555556 LA SEU D'URGELL 9771 41.617222222222225 0.6211111111111112 LLEIDA, OBSERVATORIO 9771C 41.626111111111115 0.5980555555555556 LLEIDA 9990X 42.70027777777778 0.8769444444444445 NAUT ARAN, ARTIES 9698U 42.20527777777778 0.8652777777777777 TALARN 9720X 41.650555555555556 1.1494444444444445 TÀRREGA 9981A 40.82027777777778 0.49333333333333335 ESTACIÓN DE TORTOSA (ROQUETES) 9987P 40.69916666666666 0.81 SANT JAUME D'ENVEJA

In [0]:
estacionesDF= estaciones_meteocatDF.union(estaciones_aemetDF)
display(estacionesDF)

indicativo latitud longitud nombre X4 41.3839 2.16775 Barcelona YG 42.51881 1.24244 Tírvia VY 41.25095 1.29863 Nulles YD 41.51135 0.85617 Les Borges Blanques VD 41.68939 1.20381 Els Plans de Sió WX 41.9178 0.88175 Camarasa XV 41.48311 2.07956 Sant Cugat del Vallès U1 42.30648 2.95481 Cabanes CY 41.87813 2.17873 Muntanyola D6 42.43515 3.16622 Portbou D3 41.38197 1.93564 Vallirana Z1 42.64691 0.98486 Alt Àneu X1 41.15374 0.81953 Falset VH 41.65814 0.39301 Gimenells i el Pla de la Font WK 41.81949 0.57768 Alfarràs YL 41.1227 0.96951 Riudecanyes CT 42.39811 0.74362 El Pont de Suert XR 41.31481 0.98161 Prades H1 41.58642 1.65306 Òdena UE 42.02306 3.15719 Torroella de Montgrí W4 41.36619 1.72857 La Granada X3 41.74281 0.5358 Alguaire US 40.55786 0.52329 Alcanar KP 41.73894 2.69436 Fogars de la Selva XD 41.32 0.8857 Ulldemolins XY 41.56509 0.55027 Alcarràs CJ 42.21624 1.33132 Organyà U9 40.76947 0.61635 L'Aldea YB 42.18813 2.47097 Olot D1 41.28521 0.75383 Margalef XG 41.56734 2.22619 Parets del Vallès UF 41.28832 1.90775 Begues Z3 42.46605 1.7785 Meranges CE 41.53109 1.80813 Els Hostalets de Pierola CU 42.69856 0.79397 Vielha e Mijaran WI 41.35741 0.4809 Maials UW 40.62725 0.65922 Sant Carles de la Ràpita CC 42.07398 2.20862 Orís W8 41.44237 1.15998 Blancafort C6 41.6566 0.95172 Castellnou de Seana WO 41.27151 1.46717 La Bisbal del Penedès VU 41.63286 1.91718 Rellinars WQ 42.0513 0.72952 Sant Esteve de la Sarga XQ 42.16252 0.88814 Tremp WB 41.76036 0.67022 Albesa DQ 41.30728 1.36259 Vila-rodona UI 42.26477 1.76218 Gisclareny DK 41.14677 1.41846 Torredembarra J5 42.34215 2.83423 Darnius Y9 41.31723 2.13091 El Prat de Llobregat WG 41.80104 0.64804 Algerri UJ 41.52879 1.3683 Santa Coloma de Queralt MV 42.13598 1.67295 Guixers VV 41.68129 2.02647 Sant Llorenç Savall UP 41.51773 2.37702 Cabrils XA 41.60257 1.4007 Montmaneu VC 41.24415 0.43266 Riba-roja d'Ebre Y4 42.17426 1.42775 Fígols i Alinyà Z5 42.70029 1.27201 Lladorre ZC 42.42117 2.24565 Setcases CL 41.67399 1.76796 Sant Salvador de Guardiola XT 41.98766 1.51165 Solsona X8 41.37919 2.1054 Barcelona UN 41.87449 2.92694 Cassà de la Selva WR 41.2163 0.79748 Torroja del Priorat Y5 42.22793 2.86295 Navata VE 41.48692 0.46058 Aitona UC 41.97726 2.99705 Cruïlles, Monells i Sant Sadurní de l'Heura DG 42.39848 2.15517 Queralbs XH 42.40533 1.12993 Sort V5 42.03947 2.11993 Perafita VN 41.88244 2.74262 Vilobí d'Onyar C7 41.66695 1.16234 Tàrrega YJ 41.58462 0.64172 Lleida VZ 42.38443 3.00736 Espolla WS 41.84008 2.41877 Viladrau CP 42.13924 1.03893 Isona i Conca Dellà CR 42.08032 1.9624 La Quar C9 40.71825 0.39988 Mas de Barberans VQ 41.1713 1.16774 Constantí CI 42.25839 2.36429 Sant Pau de Segúries U4 41.83094 1.84699 Castellnou de Bages MQ 41.92047 1.69584 Cardona DB 40.87287 0.71581 El Perelló D7 41.18499 0.59376 Vinebre WZ 41.20187 1.63346 Cunit WP 41.48518 1.69337 Mediona W5 42.07683 1.31489 Oliana UA 40.9091 0.76943 L'Ametlla de Mar Y7 41.31725 2.16537 El Prat de Llobregat X6 41.92173 1.02901 Artesa de Segre XE 41.10393 1.201 Tarragona XB 41.4779 1.53529 La Llacuna XZ 42.18139 3.06202 Torroella de Fluvià DO 41.80857 3.03241 Castell-Platja d'Aro V8 41.67279 0.87741 El Poal UB 42.05398 3.06195 La Tallada d'Empordà X7 41.51909 0.55314 Torres de Segre UK 41.27861 1.80479 Sant Pere de Ribes VB 41.06289 0.63517 Benissanet C8 41.67555 1.29609 Cervera WA 41.87694 1.1541 Oliola YK 41.55361 1.99005 Terrassa Z9 42.29265 1.71498 Montellà i Martinet VX 41.74761 2.30291 Tagamanent WV 42.23414 1.87485 Guardiola de Berguedà U2 42.17716 3.0968 Sant Pere Pescador D8 40.95134 0.30565 Horta de Sant Joan ZB 42.51939 1.36597 Alins WL 41.57236 1.0882 Sant Martí de Riucorb W9 42.14557 2.45451 La Vall d'en Bas WU 41.45215 2.24757 Badalona VO 42.08745 1.42792 Lladurs XO 41.93497 2.23987 Vic XC 41.47892 1.97546 Castellbisbal XK 41.77362 2.43774 Fogars de Montclús ZD 42.32211 1.89716 Alp WW 41.79423 1.9368 Artés Y6 41.04343 0.74032 Tivissa WD 41.08671 0.31884 Batea CQ 41.99546 1.02569 Vilanova de Meià XS 41.86489 2.664

In [0]:
geolocator = Nominatim(user_agent="https://maps.googleapis.com")
def getCoordinatesbyMuni(munic):
    """
    Funcion para obtener posicion lat lon dado un municipio
    """
    try:
        location = geolocator.geocode(munic)
    except Exception as e:
        print(e)
        return None
    return f"{location.raw.get('lat')} {location.raw.get('lon')}"

In [0]:
from geopy.extra.rate_limiter import RateLimiter
from tqdm import tqdm

from geopy.geocoders import Nominatim

geolocator = Nominatim(user_agent="https://maps.googleapis.com")

missing_muniPD = missing_muni.toPandas()

# Se limpia el campo municipio para una mejor precision de la api de google
missing_muniPD['municipio_clean']=missing_muniPD['municipio'].apply(lambda x: x.split(",")[0])
missing_muniPD['municipio_clean'] = missing_muniPD['municipio_clean']+ ', España'

tqdm.pandas()
missing_muniPD['location_code'] = missing_muniPD['municipio_clean'].progress_apply(getCoordinatesbyMuni)

0%| | 0/946 [00:00<?, ?it/s]
 0%| | 2/946 [00:00<02:15, 6.99it/s]
 0%| | 3/946 [00:00<04:40, 3.36it/s]
 0%| | 4/946 [00:01<05:39, 2.78it/s]
 1%| | 5/946 [00:01<06:30, 2.41it/s]
 1%| | 6/946 [00:02<06:58, 2.25it/s]
 1%| | 7/946 [00:02<07:09, 2.18it/s]
 1%| | 8/946 [00:03<07:25, 2.11it/s]
 1%| | 9/946 [00:03<07:29, 2.08it/s]
 1%| | 10/946 [00:04<09:16, 1.68it/s]
 1%| | 11/946 [00:04<07:18, 2.13it/s]
 1%|▏ | 12/946 [00:05<07:20, 2.12it/s]
 1%|▏ | 13/946 [00:05<07:27, 2.08it/s]
 1%|▏ | 14/946 [00:06<07:31, 2.06it/s]
 2%|▏ | 15/946 [00:06<08:24, 1.85it/s]
 2%|▏ | 16/946 [00:07<07:28, 2.07it/s]
 2%|▏ | 17/946 [00:07<07:34, 2.04it/s]
 2%|▏ | 18/946 [00:08<07:28, 2.07it/s]
 2%|▏ | 19/946 [00:08<07:35, 2.03it/s]
 2%|▏ | 20/946 [00:09<07:37, 2.02it/s]
 2%|▏ | 21/946 [00:09<07:39, 2.02it/s]
 2%|▏ | 22/946 [00:10<07:40, 2.01it/s]
 2%|▏ | 23/946 [00:10<07:41, 2.00it/s]
 3%|▎ | 24/946 [00:11<09:12, 1.67it/s]
 3%|▎ | 25/946 [00:11<07:09, 2.14it/s]
 3%|▎ | 26/946 [00:12<07:33, 2.03it/s]
 3%|▎ | 27/946 [00:13<08:28, 1.81it/s]
 3%|▎ | 28/946 [00:13<07:06, 2.15it/s]
 3%|▎ | 29/946 [00:13<07:15, 2.11it/s]
 3%|▎ | 30/946 [00:14<07:15, 2.10it/s]
 3%|▎ | 31/946 [00:14<07:25, 2.05it/s]
 3%|▎ | 32/946 [00:15<07:24, 2.06it/s]
 3%|▎ | 33/946 [00:15<07:32, 2.02it/s]
 4%|▎ | 34/946 [00:16<07:34, 2.01it/s]
 4%|▎ | 35/946 [00:16<08:15, 1.84it/s]
 4%|▍ | 36/946 [00:17<07:28, 2.03it/s]
 4%|▍ | 37/946 [00:17<07:35, 2.00it/s]
 4%|▍ | 38/946 [00:18<07:23, 2.05it/s]
 4%|▍ | 39/946 [00:18<07:24, 2.04it/s]
 4%|▍ | 40/946 [00:19<07:26, 2.03it/s]
 4%|▍ | 41/946 [00:19<07:33, 1.99it/s]
 4%|▍ | 42/946 [00:20<07:29, 2.01it/s]
 5%|▍ | 43/946 [00:20<07:26, 2.02it/s]
 5%|▍ | 44/946 [00:21<07:28, 2.01it/s]
 5%|▍ | 45/946 [00:21<07:42, 1.95it/s]
 5%|▍ | 46/946 [00:22<07:27, 2.01it/s]
 5%|▍ | 47/946 [00:22<07:23, 2.03it/s]
 5%|▌ | 48/946 [00:23<07:21, 2.03it/s]
 5%|▌ | 49/946 [00:23<07:27, 2.00it/s]
 5%|▌ | 50/946 [00:24<07:31, 1.98it/s]
 5%|▌ | 51/946 [00:24<07:26, 2.00it/s]
 5%|▌ | 52/946 [00:25<07:22, 2.02it/s]
 6%|▌ | 53/946 [00:25<07:23, 2.01it/s]
 6%|▌ | 54/946 [00:26<07:25, 2.00it/s]
 6%|▌ | 55/946 [00:26<07:24, 2.00it/s]
 6%|▌ | 56/946 [00:27<07:23, 2.01it/s]
 6%|▌ | 57/946 [00:27<07:27, 1.99it/s]
 6%|▌ | 58/946 [00:28<07:19, 2.02it/s]
 6%|▌ | 59/946 [00:28<07:26, 1.99it/s]
 6%|▋ | 60/946 [00:29<07:26, 1.99it/s]
 6%|▋ | 61/946 [00:29<07:20, 2.01it/s]
 7%|▋ | 62/946 [00:30<07:22, 2.00it/s]
 7%|▋ | 63/946 [00:30<07:20, 2.01it/s]
 7%|▋ | 64/946 [00:31<07:20, 2.00it/s]
 7%|▋ | 65/946 [00:31<07:22, 1.99it/s]
 7%|▋ | 66/946 [00:32<07:19, 2.00it/s]
 7%|▋ | 67/946 [00:32<07:20, 2.00it/s]
 7%|▋ | 68/946 [00:33<07:27, 1.96it/s]
 7%|▋ | 69/946 [00:33<07:25, 1.97it/s]
 7%|▋ | 70/946 [00:34<07:14, 2.02it/s]
 8%|▊ | 71/946 [00:34<07:11, 2.03it/s]
 8%|▊ | 72/946 [00:35<07:20, 1.98it/s]
 8%|▊ | 73/946 [00:35<07:11, 2.02it/s]
 8%|▊ | 74/946 [00:36<07:15, 2.00it/s]
 8%|▊ | 75/946 [00:36<07:12, 2.01it/s]
 8%|▊ | 76/946 [00:37<07:11, 2.02it/s]
 8%|▊ | 77/946 [00:37<07:16, 1.99it/s]
 8%|▊ | 78/946 [00:38<07:17, 1.99it/s]
 8%|▊ | 79/946 [00:38<07:14, 2.00it/s]
 8%|▊ | 80/946 [00:39<07:14, 1.99it/s]
 9%|▊ | 81/946 [00:40<08:22, 1.72it/s]
 9%|▊ | 82/946 [00:40<07:02, 2.05it/s]
 9%|▉ | 83/946 [00:40<06:52, 2.09it/s]
 9%|▉ | 84/946 [00:41<06:58, 2.06it/s]
 9%|▉ | 85/946 [00:41<07:03, 2.03it/s]
 9%|▉ | 86/946 [00:42<06:58, 2.05it/s]
 9%|▉ | 87/946 [00:42<07:05, 2.02it/s]
 9%|▉ | 88/946 [00:43<07:04, 2.02it/s]
 9%|▉ | 89/946 [00:43<07:08, 2.00it/s]
 10%|▉ | 90/946 [00:44<07:20, 1.94it/s]
 10%|▉ | 91/946 [00:44<07:03, 2.02it/s]
 10%|▉ | 92/946 [00:45<07:01, 2.03it/s]
 10%|▉ | 93/946 [00:45<07:03, 2.01it/s]
 10%|▉ | 94/946 [00:46<07:07, 1.99it/s]
 10%|█ | 95/946 [00:46<07:00, 2.02it/s]
 10%|█ | 96/946 [00:47<07:07, 1.99it/s]
 10%|█ | 97/946 [00:47<06:59, 2.02it/s]
 10%|█ | 98/946 [00:48<07:02, 2.01it/s]
 10%|█ | 99/946 [00:48<07:06, 1.99it/s]
 11%|█ | 100/946 [00:49<07:02, 2.00it/s]
 11%|█ | 101/946 [00:49<07:02, 2.00it/s]
 11%|█ | 102/946 [00:50<07:02, 2.00it/s]
 11%|█ | 103/946 [00:51<09:06, 1.5

In [0]:
missing_muniPD[missing_muniPD['location_code'].isna()]

Out[16]:

codigo_municipio 
 municipio 
 municipio_clean 
 location_code

In [0]:
#Algunos municipios no se han  podido obtener. Se mapean manualmente
missing_muniPD.loc[missing_muniPD['municipio']=='Port de la Selva, el','location_code']='42.3373159 3.2047232'
# missing_muniPD.loc[missing_muniPD['municipio']=='Pont de Bar, el','location_code'] ='42.3721732 1.6047661'
missing_muniPD.loc[missing_muniPD['municipio']=="Vall d'en Bas, la",'location_code'] ='42.1388600 2.4410500'
# missing_muniPD.loc[missing_muniPD['municipio']=="Ametlla de Mar, l'",'location_code'] ='40.8838957 0.8024645'

In [0]:
missing_muniDF=spark.createDataFrame(missing_muniPD).select(["municipio","location_code","codigo_municipio"])

missing_muniDF= missing_muniDF.withColumn("municipio_lat", F.split("location_code"," ").getItem(0)) \
.withColumn("municipio_lon", F.split("location_code"," ").getItem(1))
display(missing_muniDF)

municipio location_code codigo_municipio municipio_lat municipio_lon Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 Vila-seca 41.1109268 1.144961 431711 41.1109268 1.144961 Alt Àneu 42.69035515 1.0624832582072457 250241 42.69035515 1.0624832582072457 Sant Esteve de Palautordera 41.7040768 2.4343426 82074 41.7040768 2.4343426 Estany, l' 41.8858007 0.3685293 80790 41.8858007 0.3685293 Cabanyes, les 39.2908201 -0.5698989 80272 39.2908201 -0.5698989 Sant Martí Sarroca 41.3854214 1.6108158 82270 41.3854214 1.6108158 Siurana 42.2094291 3.0061564 170524 42.2094291 3.0061564 Nou de Gaià, la 41.1822282 1.3738261 430976 41.1822282 1.3738261 Malla 41.887016 2.2354564 81115 41.887016 2.2354564 Selva de Mar, la 41.4080147 2.2091094 171886 41.4080147 2.2091094 Móra d'Ebre 41.0893373 0.6412887 430939 41.0893373 0.6412887 Agullana 42.3938409 2.8469168 170010 42.3938409 2.8469168 Pobla de Segur, la 42.2473665 0.9672725 251713 42.2473665 0.9672725 Arenys de Mar 41.5797031 2.5491562 80060 41.5797031 2.5491562 Badalona 41.4493539 2.248254 80155 41.4493539 2.248254 Vilalba dels Arcs 41.1203999 0.4097144 431750 41.1203999 0.4097144 Prat de Comte 40.9829428 0.4057612 431172 40.9829428 0.4057612 Clariana de Cardener 41.9337762 1.6255629 250753 41.9337762 1.6255629 Sort 42.4114509 1.1298362 252094 42.4114509 1.1298362 Mediona 41.47739425 1.635018453270908 81228 41.47739425 1.635018453270908 Verdú 41.6102775 1.143512 252424 41.6102775 1.143512 Coma i la Pedra, la 42.17449855 1.5817023533792938 251636 42.17449855 1.5817023533792938 Prats de Lluçanès 42.0085421 2.0298689 81712 42.0085421 2.0298689 Masies de Voltregà, les 42.0267288 2.2347147 81173 42.0267288 2.2347147 Palma d'Ebre, la 41.2833974 0.6656065 430995 41.2833974 0.6656065 Bell-lloc d'Urgell 41.6321157 0.7769267 250485 41.6321157 0.7769267 Santa Eugènia de Berga 41.9000058 2.283319 82460 41.9000058 2.283319 Estaràs 41.6918459 1.3781492 250851 41.6918459 1.3781492 Torre de Fontaubella, la 41.1247206 0.8656245 431515 41.1247206 0.8656245 Castelló de Farfanya 41.8195348 0.7284574 250693 41.8195348 0.7284574 Sant Llorenç d'Hortons 41.4681755 1.8253181 82227 41.4681755 1.8253181 Vilafranca del Penedès 41.3463825 1.6995213 83054 41.3463825 1.6995213 Valls d'Aguilar, les 42.31000255 1.245256225918244 259062 42.31000255 1.245256225918244 Llosses, les 42.1506279 2.1166718 170963 42.1506279 2.1166718 Tàrrega 41.6472761 1.1409115 252173 41.6472761 1.1409115 Torelló 42.0478304 2.2643391 82858 42.0478304 2.2643391 Alella 41.4952867 2.2942758 80039 41.4952867 2.2942758 Castellar de n'Hug 42.2826009 2.0165927 80522 42.2826009 2.0165927 Sant Pau de Segúries 42.2633408 2.366055 171772 42.2633408 2.366055 Berga 42.1011456 1.8454758 80229 42.1011456 1.8454758 Montoliu de Segarra 41.5901454 1.2695473 251410 41.5901454 1.2695473 Foradada 41.877024 1.0143414 250946 41.877024 1.0143414 Llers 42.2956869 2.9114592 170932 42.2956869 2.9114592 Marçà 41.1281164 0.8021678 430764 41.1281164 0.8021678 Sant Martí d'Albars 42.0281258 2.0742512 82251 42.0281258 2.0742512 Masdenverge 40.7150044 0.5311135 430786 40.7150044 0.5311135 Castellolí 41.5981563 1.7002468 80636 41.5981563 1.7002468 Olivella 41.3103129 1.8114946 81483 41.3103129 1.8114946 Aiguaviva 41.9382389 2.7624155 170025 41.9382389 2.7624155 Capçanes 41.1003901 0.7810956 430402 41.1003901 0.7810956 Maçanet de Cabrenys 42.3842351 2.7569154 171024 42.3842351 2.7569154 Vilaller 42.4759263 0.7165615 252458 42.4759263 0.7165615 Cervelló 41.3964186 1.9585925 80689 41.3964186 1.9585925 Montagut i Oix 42.2309257 2.5967415 171096 42.2309257 2.5967415 Sant Cugat del Vallès 41.4728432 2.0817809 82055 41.4728432 2.0817809 Pira 41.4236202 1.2031023 431074 41.4236202 1.2031023 Farrera 42.4969344 1.27255 250899 42.4969344 1.27255 Santa Maria de Merlès 42.0007572 1.9778885 82554 42.0007572 1.9778885 Esparreguera 41.5393249 1.8694549 80765 41.5393249 1.8694549 Sant Joan de Vilatorrada 41.7422073 1.803625 82188 41.7422073 1.803625 Maià de Montcal 42.2221769 2.7431144 1709

In [0]:
# Se ejecuta una crossjoin para tener todas la posibilidades de estaciones y
#municipios con el fin de calcular la distancia entre todos ellos y finalmente obtener la estacion más cercana

munic_estaciones_convinations = missing_muniDF.crossJoin(estacionesDF)
munic_estaciones_convinations.count()

Out[18]: 198660

In [0]:
from geopy.distance import geodesic
from pyspark.sql.types import FloatType

@F.udf(returnType=FloatType())
def geodesic_udf(a, b):
    """
    Funcion para calcular la distancia entre dos puntos de coordinadas
    """
    return geodesic(a, b).m

estacion_munic_distancia = munic_estaciones_convinations.withColumn('estacion_muni_distancia_m', geodesic_udf(F.array("municipio_lat", "municipio_lon"), F.array("latitud", "longitud")))
display(estacion_munic_distancia)

municipio location_code codigo_municipio municipio_lat municipio_lon indicativo latitud longitud nombre estacion_muni_distancia_m Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 X4 41.3839 2.16775 Barcelona 75321.96 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 YG 42.51881 1.24244 Tírvia 114870.95 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 VY 41.25095 1.29863 Nulles 129723.266 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 YD 41.51135 0.85617 Les Borges Blanques 145431.02 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 VD 41.68939 1.20381 Els Plans de Sió 111165.33 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 WX 41.9178 0.88175 Camarasa 131795.7 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 XV 41.48311 2.07956 Sant Cugat del Vallès 68141.54 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 U1 42.30648 2.95481 Cabanes 51164.848 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 CY 41.87813 2.17873 Muntanyola 28786.14 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 D6 42.43515 3.16622 Portbou 73684.72 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 D3 41.38197 1.93564 Vallirana 83871.35 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 Z1 42.64691 0.98486 Alt Àneu 140248.5 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 X1 41.15374 0.81953 Falset 167872.23 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 VH 41.65814 0.39301 Gimenells i el Pla de la Font 176859.81 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 WK 41.81949 0.57768 Alfarràs 158230.34 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 YL 41.1227 0.96951 Riudecanyes 159973.03 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 CT 42.39811 0.74362 El Pont de Suert 148112.2 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 XR 41.31481 0.98161 Prades 146553.0 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 H1 41.58642 1.65306 Òdena 83203.54 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 UE 42.02306 3.15719 Torroella de Montgrí 57294.746 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 W4 41.36619 1.72857 La Granada 95422.15 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 X3 41.74281 0.5358 Alguaire 163179.31 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 US 40.55786 0.52329 Alcanar 230181.62 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 KP 41.73894 2.69436 Fogars de la Selva 36959.77 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 XD 41.32 0.8857 Ulldemolins 153047.0 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 XY 41.56509 0.55027 Alcarràs 167146.7 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 CJ 42.21624 1.33132 Organyà 96172.75 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 U9 40.76947 0.61635 L'Aldea 208159.39 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 YB 42.18813 2.47097 Olot 18199.547 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 D1 41.28521 0.75383 Margalef 164508.77 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 XG 41.56734 2.22619 Parets del Vallès 54512.695 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 UF 41.28832 1.90775 Begues 94020.69 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 Z3 42.46605 1.7785 Meranges 74974.08 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 CE 41.53109 1.80813 Els Hostalets de Pierola 77376.74 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 CU 42.69856 0.79397 Vielha e Mijaran 156737.83 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 WI 41.35741 0.4809 Maials 181049.66 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 UW 40.62725 0.65922 Sant Carles de la Ràpita 216645.34 Rupit i Pruit 42.024334 2.4653919 89019 42.024334 2.4653919 CC 42.07398 2.20862 Orís 21960.959 Rupit i Pruit 42.02

In [0]:
#Calculamos un ranking de estaciones mas cercanas a mas lejanas
# La estacion con row_number igual a 1 sera la mas cercana
windowSpec  = Window.partitionBy("municipio").orderBy("estacion_muni_distancia_m")

estacion_munic_distancia=estacion_munic_distancia.withColumn("row_number",row_number().over(windowSpec))
display(estacion_munic_distancia)

municipio location_code codigo_municipio municipio_lat municipio_lon indicativo latitud longitud nombre estacion_muni_distancia_m row_number Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 XC 41.47892 1.97546 Castellbisbal 7646.6675 1 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 CE 41.53109 1.80813 Els Hostalets de Pierola 7957.4097 2 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 D2 41.59252 1.915 Vacarisses 8073.4136 3 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 YK 41.55361 1.99005 Terrassa 8188.393 4 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 WN 41.59539 1.83751 Monistrol de Montserrat 9929.583 5 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 VU 41.63286 1.91718 Rellinars 12545.819 6 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 WY 41.43386 1.79429 Sant Sadurní d'Anoia 13192.308 7 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 XF 41.56568 2.06952 Sabadell 14821.011 8 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 XV 41.48311 2.07956 Sant Cugat del Vallès 15361.012 9 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 D3 41.38197 1.93564 Vallirana 15628.157 10 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 0229I 41.52361111111111 2.1030555555555557 SABADELL AEROPUERTO 16750.924 11 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 WP 41.48518 1.69337 Mediona 17887.592 12 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 CL 41.67399 1.76796 Sant Salvador de Guardiola 20407.6 13 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 VV 41.68129 2.02647 Sant Llorenç Savall 20641.957 14 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 R1 41.70637 1.87276 El Pont de Vilomara i Rocafort 20797.133 15 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 D5 41.41843 2.12388 Barcelona 21694.186 16 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 0200E 41.41833333333333 2.1241666666666665 BARCELONA, FABRA 21720.227 17 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 H1 41.58642 1.65306 Òdena 22055.578 18 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 W4 41.36619 1.72857 La Granada 22462.557 19 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 0149X 41.72 1.840277777777778 MANRESA 22760.518 20 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 X8 41.37919 2.1054 Barcelona 23104.613 21 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 X9 41.61265 2.16836 Caldes de Montbui 24432.504 22 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 XL 41.34045 2.08022 El Prat de Llobregat 24910.115 23 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 DI 41.43292 1.62386 Font-rubí 25215.516 24 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 UF 41.28832 1.90775 Begues 25784.223 25 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 X4 41.3839 2.16775 Barcelona 26861.371 26 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 UG 41.29928 2.03787 Viladecans 27048.518 27 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 XG 41.56734 2.22619 Parets del Vallès 27513.34 28 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 U3 41.37491 1.63032 Sant Martí Sarroca 27896.805 29 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 X2 41.38943 2.18847 Barcelona 27982.152 30 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 UK 41.27861 1.80479 Sant Pere de Ribes 28072.014 31 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 0201D 41.39055555555556 2.2 BARCELONA 28746.264 32 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 0076 41.29277777777778 2.0700000000000003 BARCELONA AEROPUERTO 28908.621 33 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 Y9 41.31723 2.13091 El Prat de Llobregat 29568.4 34 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 WU 41.45215 2.24757 Badalona 29806.67 35 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 XU 41.28801 1.72195 Canyelles 29901.734 36 Abrera 41.5204464 1.9024126 80018 41.5204464 1.9024126 WW 41.79423 1.9368 Artés 30542.822

In [0]:
estacion_munic_distancia= estacion_munic_distancia.filter(F.col("row_number")=="1").select(["codigo_municipio","municipio","indicativo","nombre","row_number"]).withColumnRenamed("nombre","nombre_estacion")

In [0]:
display(estacion_munic_distancia)

codigo_municipio municipio indicativo nombre_estacion row_number 80018 Abrera XC Castellbisbal 1 250045 Alamús, els XM Els Alamús 1 170046 Albons UB La Tallada d'Empordà 1 430081 Alfara de Carles U7 Aldover 1 430094 Alforja MR Cornudella de Montsant 1 250143 Alfés X7 Torres de Segre 1 250169 Alguaire X3 Alguaire 1 250215 Almenar WK Alfarràs 1 250241 Alt Àneu Z1 Alt Àneu 1 430136 Ametlla de Mar, l' UA L'Ametlla de Mar 1 430141 Amposta UU Amposta 1 170097 Arbúcies XK Fogars de Montclús 1 80076 Arenys de Munt 0252D ARENYS DE MAR 1 80095 Argentona UP Cabrils 1 80082 Argençola XA Montmaneu 1 430189 Arnes D8 Horta de Sant Joan 1 250313 Arres Z6 Arres 1 250328 Arsèguel 9619 LA SEU D'URGELL 1 250371 Avellanes i Santa Linya, les UY Os de Balaguer 1 80137 Avinyonet del Penedès W4 La Granada 1 250404 Balaguer V1 Vallfogona de Balaguer 1 80207 Begues UF Begues 1 170139 Begur UE Torroella de Montgrí 1 250463 Belianes WL Sant Martí de Riucorb 1 250485 Bell-lloc d'Urgell XM Els Alamús 1 250479 Bellcaire d'Urgell V1 Vallfogona de Balaguer 1 80214 Bellprat UJ Santa Coloma de Queralt 1 430249 Bellvei D9 El Vendrell 1 170216 Beuda Y5 Navata 1 430271 Bisbal de Falset, la D1 Margalef 1 430290 Blancafort W8 Blancafort 1 170237 Blanes WT Malgrat de Mar 1 170293 Boadella i les Escaules J5 Darnius 1 250582 Borges Blanques, les YD Les Borges Blanques 1 430311 Borges del Camp, les YL Riudecanyes 1 170268 Borrassà Y5 Navata 1 250595 Bossòst Z6 Arres 1 430332 Botarell YL Riudecanyes 1 170280 Brunyola i Sant Martí Sapresa XS Santa Coloma de Farners 1 430363 Cabra del Camp W8 Blancafort 1 80288 Cabrera d'Anoia WP Mediona 1 80305 Cabrils UP Cabrils 1 170329 Cadaqués D4 Roses 1 80312 Calaf YH Pujalt 1 80348 Calders WW Artés 1 170335 Caldes de Malavella VN Vilobí d'Onyar 1 80370 Calldetenes XO Vic 1 430385 Cambrils U6 Vinyols i els Arcs 1 80399 Campins XK Fogars de Montclús 1 170353 Camós 0372C PORQUERES 1 80410 Canovelles KX La Roca del Vallès 1 170412 Cantallops VZ Espolla 1 80431 Canyelles XU Canyelles 1 80446 Capellades WP Mediona 1 80459 Capolat WM Berga 1 430402 Capçanes WJ El Masroig 1 170486 Castell-Platja d'Aro DO Castell-Platja d'Aro 1 80538 Castellbell i el Vilar WN Monistrol de Montserrat 1 80556 Castellcir VX Tagamanent 1 80608 Castellfollit de Riubregós YH Pujalt 1 80594 Castellfollit del Boix CL Sant Salvador de Guardiola 1 80620 Castellnou de Bages U4 Castellnou de Bages 1 80654 Castellví de la Marca U3 Sant Martí Sarroca 1 430430 Catllar, el DK Torredembarra 1 171899 Cellera de Ter, la DN Anglès 1 82687 Cercs WM Berga 1 170502 Cervià de Ter DJ Banyoles 1 250735 Cervià de les Garrigues YD Les Borges Blanques 1 250753 Clariana de Cardener MQ Cardona 1 170558 Colomers UB La Tallada d'Empordà 1 80713 Copons XA Montmaneu 1 250788 Corbins VM Vilanova de Segrià 1 170561 Cornellà del Terri DJ Banyoles 1 170577 Corçà DF La Bisbal d'Empordà 1 430509 Creixell DK Torredembarra 1 80749 Cubelles WZ Cunit 1 430516 Cunit WZ Cunit 1 170600 Darnius J5 Darnius 1 80765 Esparreguera CE Els Hostalets de Pierola 1 430542 Espluga de Francolí, l' CW L'Espluga de Francolí 1 250827 Espot Z7 Espot 1 250864 Esterri d'Àneu Z1 Alt Àneu 1 430555 Falset X1 Falset 1 170059 Far d'Empordà, el U1 Cabanes 1 430568 Fatarella, la VA Ascó 1 430580 Figuera, la VP Pinós 1 170681 Foixà UB La Tallada d'Empordà 1 80847 Fonollosa VP Pinós 1 170715 Fontcoberta DJ Banyoles 1 430614 Forès W8 Blancafort 1 430629 Freginals UU Amposta 1 250962 Fuliola, la XX Tornabous 1 259084 Fígols i Alinyà Y4 Fígols i Alinyà 1 80879 Gallifa VV Sant Llorenç Savall 1 170754 Garrigàs Y5 Navata 1 80919 Gelida WY Sant Sadurní d'Anoia 1 259123 Gimenells i el Pla de la Font VH Gimenells i el Pla de la Font 1 170806 Gombrèn 0324A RIPOLL 1 251039 Granyanella VD Els Plans de Sió 1 430691 Gratallops WR Torroja del Priorat 1 80977 Gualba XK Fogars de Montclús 1 80996 Guardiola de Berguedà WV Guardiola de Berguedà 1 251095 Guimerà WL Sant Martí de Riucorb 1 251001 Gósol Z9 Montellà i Martinet 1 81017 Hospitalet de Llobreg

In [0]:
estacion_munic_distancia.write.mode("overwrite").parquet(f"/mnt/IncendiosForestalesCAT/prep/meteocat/estaciones_cercanas_municipio/")